frame/contracts/src/lib.rs

```diff
@@ -439,8 +456,19 @@ decl_module! {
 		/// The maximum size of a storage value in bytes. A reasonable default is 16 KiB.
 		const MaxValueSize: u32 = T::MaxValueSize::get();
 
+		/// The maximum number of tries that can be queued for deletion.
+		const DeletionQueueDepth: u32 = T::DeletionQueueDepth::get();
+
+		/// The maximum amount of weight that can be consumed per block for lazy trie removal.
+		const DeletionWeightLimit: Weight = T::DeletionWeightLimit::get();
+
 		fn deposit_event() = default;
 
+		fn on_initialize() -> Weight {
+			Storage::<T>::process_deletion_queue_batch(T::DeletionWeightLimit::get())
```

Comment:
maybe it should take the min of DeletionWeightLimit and the still available weight, so that on runtime upgrade we don't take even more weight

Reply:
Yes this is a good idea. Also it got me thinking. The lazy storage removal is actually optional and not some mandatory thing that must run every block. This means that `on_initialize` might be the wrong place to put it. It should be some operational extrinsic that is submitted only when it fits the block. That would solve the problem of it impacting the `AVERAGE_ON_INITIALIZE_RATIO`. How would I do this? @shawntabrizi

Reply:
maybe using offchain worker ?

Reply:
Yes that could be a long term solution. I think for now we can get away with tuning the weight limit down.

Reply:
> Is there a function for that or do I need to calculate that myself?

frame_system expose the function `block_weight` to get the currently consummed weight, and then you should use `T::BlockWeights::get()` which returns `limits::BlockWeights`, maybe add a function to `limits::BlockWeights` or just substract consummed weight to `block_weights.max_block` (or to `block_weights.get(normal).max_total`)

Reply:
Thanks I will look into it.

Reply:
I added the new code that should prevent the deletion in case of an already full block.

Reply:
Yes look good to me 👍